elastic/kibana
# [Security Solution][Detections] Updates Rule Details to show all historical alerts for a given Rule

## Summary

Resolves Part II of https://github.com/elastic/kibana/issues/118935 which fixes a bug where pre-8.x alerts are not displayed on the `Rule Details` page if the [Rule SO ID had been re-generated](https://www.elastic.co/guide/en/kibana/master/sharing-saved-objects.html#sharing-saved-objects-faq-changing-object-ids) as part of the upgrade to 8.x. As discussed with the team and relevant product folks, we're resolving this issue by updating the alerts query to filter by `ruleId` instead of the Rule SO ID, which will result in a slight workflow change where _all alerts_ for a given `ruleId` will be displayed rather than just the latest instance -- i.e. create pre-package rule, generate alerts, delete/re-create pre-package rule, generate alerts, alerts are displayed from both Rule instances instead of just the most recent. This also updates the exceptions workflow such that if the user specifies `close all associated alerts` when creating a Rule Exception, it will now close all alerts matching the exception _for all historical instances_ of the rule, maintaining consistent behavior when managing alerts from a rule.

The Previous _Rule Deleted_ functionality stays mostly the same -- if you navigate to `Rule Details` from an older alert whose rule has been deleted, the Rule Details page will still re-construct the rule's details from the alert, and now show all alerts (as above) instead of just that instances' alerts.

Also cleaned up a lot of duplicate/redundant code over in `default_config.tsx` that was leftover from the `RuleRegsitry` feature flag. Now condensed back to one function each for generating filters/columns/config.

## Test Instructions:

Clear `es-data`, and start `7.16` ES
```
rm -rf $DEV_HOME/es-data
es snapshot --license trial -E xpack.security.authc.api_key.enabled=true -E path.data=$DEV_HOME/es-data --version=7.16.0
```
Navigate to `kibana-7.16`, checkout and `yarn start`
Login, create a rule, then inject some data
```
node x-pack/plugins/security_solution/scripts/endpoint/resolver_generator.js --node http://elastic:changeme@127.0.0.1:9200 --kibana http://elastic:changeme@localhost:5601/kbn --numHosts=5 --numDocs=2  --fleet
```
Verify Alerts, kill kibana, kill es, then start es `8.1.0` (run from `kibana-main` or use `--version=8.1.0`)
```
es snapshot --license trial -E xpack.security.authc.api_key.enabled=true -E path.data=$DEV_HOME/es-data
```
Navigate to `kibana-main`, checkout and `yarn start`
Login, navigate to previous rule (toggling enable/disable to bypass [this bug](https://github.com/elastic/security-team/issues/2344)), then inject some data (previous `resolver_generator` data-gen command should've created enough future events, but if not re-run w/o `--fleet` flag)

### Verify `7.16` and `8.1` alerts are visible in both the Alerts Histogram and Alerts Table:

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/144160208-3acbf215-3270-4e38-b723-0a59173eab2f.png" />
</p>


Note, `Severity`/`Risk Score` will not display for pre-8.x alerts see [this comment](https://github.com/elastic/kibana/pull/119921#issuecomment-983130809) for details.

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/144145912-1ffba09e-016b-4bcf-8b7f-93c29ad72988.png" />
</p>

### Verify `7.16` and `8.1` alerts are closed when adding a Rule Exception (that of course matches all alerts ;) and selecting `Close all alerts that match this exception`.

<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/144957832-3ce7af77-6102-4fe1-914f-0f87b9a2983e.gif" />
</p>



Note: Initial approach was to plumb through the `legacyId` through the Read Rules route (https://github.com/elastic/kibana/pull/119031), so we can always revisit that method if necessary.

### Checklist
Delete any items that are not applicable to this PR.
